### ISSUE TITLE
Saving/loading models with tf.keras.layers.experimental.preprocessing.Hashing layer produces inconsistent results

### ISSUE THREAD

**Issue by DavidPal** [keras-team/keras#15061]
**System information**.
Colab notebook: https://colab.research.google.com/drive/1J0u_vHYGUfMLxFxiT6AtVqS415KlWk7O

**Describe the problem**.
When a keras model contains `tf.keras.layers.experimental.preprocessing.Hashing` layer without explicitly setting
`salt` parameter (i.e. `salt=None`) then saving and loading the model from disk again produces inconsistent results.
Namely, the hash functions used before and after saving&loading the model are different!

According to documentation:
https://www.tensorflow.org/api_docs/python/tf/keras/layers/experimental/preprocessing/Hashing
```
This layer uses FarmHash64 by default, which provides a consistent hashed output across different platforms 
and is stable across invocations, regardless of device and context, by mixing the input bits thoroughly.
```
This seems to suggest that setting `salt=None` is legal and saving&loading the model should produce consistent results.

My guess is that the code that serializes and deserializes the Hashing layer does not handle the case `salt=None` correctly.

**Describe the current behavior**.
When a keras model contains `tf.keras.layers.experimental.preprocessing.Hashing` layer without explicitly setting
`salt` parameter (i.e. `salt=None`) then saving and loading the model from disk again produces inconsistent results.
Namely, the hash functions used before and after saving&loading the model are different.

**Describe the expected behavior**.
When a keras model contains `tf.keras.layers.experimental.preprocessing.Hashing` layer without explicitly setting
`salt` parameter (i.e. `salt=None`) then saving and loading the model from disk again should produce use the *same hash function*.

**Standalone code to reproduce the issue**.

Colab notebook: https://colab.research.google.com/drive/1J0u_vHYGUfMLxFxiT6AtVqS415KlWk7O

```python
# This piece of code demonstrates 
# a bug in tf.keras.layers.experimental.preprocessing.Hashing layer.
#
# When a Keras model contains a tf.keras.layers.experimental.preprocessing.Hashing
# layer without salt (i.e. salt=None) then saving the model disk and 
# loading the model will create incosistent result.
#
# My guess is that the code that serializes and deserializes the Hashing layer
# does not handle the case `salt=None` correctly.
#
# Setting salt to any integer value fixes the problem. However, according to the
# documentation (https://www.tensorflow.org/api_docs/python/tf/keras/layers/experimental/preprocessing/Hashing)
# using `salt=None` is perfectly legal and should be produce consistent results:
#
#   "This layer uses FarmHash64 by default, which provides a consistent hashed 
#   output across different platforms and is stable across invocations, 
#   regardless of device and context, by mixing the input bits thoroughly."
#
# The bug was discovered on August 3, 2021 by:
#   David Pal (davidko.pal@gmail.com, david.pal@instacart.com) 
#   Peng Qi (peng.qi@instacart.com)

import tensorflow as tf

# Create an input layer.
input_layer = tf.keras.layers.Input(shape=(1,),
                                    name="input_text",
                                    dtype=tf.string)

# Create a hashing layer WITHOUT salt.
hashing_layer = tf.keras.layers.experimental.preprocessing.Hashing(num_bins=100)

# Connect input layer to the hashing layer.
output = hashing_layer(input_layer)

# Create a trivial Keras model.
model = tf.keras.Model(inputs=input_layer, outputs=output, name="Test")

# Make some "predictions" using the model.
print(model.predict(['hello', 'world']))

# Save the model to disk.
model.save("/tmp")

# Load the model from disk.
loaded_model = tf.keras.models.load_model("/tmp")

# Make some "predictions" using the model.
print(loaded_model.predict(['hello', 'world']))
```

**Comment by rmothukuru** [comment]
@DavidPal,
Do you mean to say that the bug is that **`first Print`** is returning **`0x7f312a5584d0`** and **`second print`** is returning **`0x7f312a4f27a0`**, but they are supposed to be the same. Is my understanding correct?

**Comment by peng-2020** [comment]
@rmothukuru , no. Please ignore the warnings and the memory address. 
The output of the model is
[[96]
 [98]]
and the output of the loaded model is
[[74]
 [47]]

**Comment by DavidPal** [comment]
Let me be very explicit. The first print statement prints
[[96]
[98]]
These are the hash codes for "hello" and "world" respectively.

The second print statement prints
[[74]
[47]]
These are again hash codes for "hello" and "world" respectively.

The bug is that [[96], [98]] != [[74], [47]]. The expected behavior is that the two lists would be the same.

**Comment by mattdangerw** [comment]
I believe this was a bug in 2.5, that is now fixed. Can you check if using 2.6 fixes the issue (colab should use it by default now).

Thanks for filing!

**Comment by mattdangerw** [comment]
Closing, this should be fixed as of 2.6. Please reopen if that is not the case!
